plaid/src/data/websocket/selector.rs

```diff
@@ -0,0 +1,141 @@
+    /// has not yet passed, the process will wait until it has before retrying.
+    uris: BinaryHeap<UriEntry>,
+    /// The initial duration to wait before retrying a connection.
+    initial_retry_after: Duration,
+    /// The maximum duration to wait before retrying a connection.
+    max_retry_after: Duration,
+}
+
+impl Ord for UriEntry {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        // Reverse the comparison to make the BinaryHeap a min-heap
+        self.next_attempt.cmp(&other.next_attempt).reverse()
+    }
+}
+
+impl PartialOrd for UriEntry {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl UriSelector {
+    /// Creates a new `UriSelector` with the given list of URIs, initial retry duration, and maximum retry duration.
+    ///
+    /// # Arguments
+    ///
+    /// * `uris` - A vector of `Uri` objects representing the URIs to manage.
+    /// * `initial_retry_after` - A `Duration` specifying the initial time to wait before retrying a failed URI.
+    /// * `max_retry_after` - A `Duration` specifying the maximum time to wait before retrying a failed URI.
+    ///
+    /// # Returns
+    ///
+    /// A `UriSelector` instance with the given URIs and retry durations.
+    pub fn new(
+        uris: HashMap<String, Uri>,
+        initial_retry_after: Duration,
+        max_retry_after: Duration,
+    ) -> Self {
+        let now = Instant::now();
+        UriSelector {
+            uris: BinaryHeap::from(
+                uris.into_iter()
+                    .map(|(name, uri)| UriEntry {
+                        name,
+                        uri,
+                        backoff_duration: initial_retry_after,
+                        next_attempt: now,
+                    })
+                    .collect::<Vec<_>>(),
+            ),
+            initial_retry_after,
+            max_retry_after,
+        }
+    }
+
+    /// Selects the next URI to use, prioritizing URIs with the shortest backoff duration.
+    ///
+    /// This function sorts the URIs by their next attempt time,
+    /// and returns a tuple containing the name and URI of the entry with the shortest backoff duration that is ready for the next attempt.
+    /// If no URIs are ready, it selects and returns the name and URI of the entry with the earliest next attempt time.
+    ///
+    /// # Returns
+    ///
+    /// A tuple `(String, Uri)` where the `String` is the name of the URI and the `Uri` is the next URI to use.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the URI list is empty, although in practice this should never be possible.
+    pub async fn next_uri(&self) -> Option<(String, Uri)> {
+        // Select the URI with the shortest backoff duration that is ready for the next attempt
+        let uri = self.uris.peek()?;
+        let now = Instant::now();
+
+        if now >= uri.next_attempt {
```

Comment:
I would flip this so you know if you have to sleep, then you can crush the returns together.